File: trie.py
"""Trie data structure"""

class Node(object):

    def __init__(self):
        self.edges = {}
        self.word = None

    def __getitem__(self, key):
        return self.edges[key]

    def __setitem__(self, key, value):
        self.edges[key] = value

    # TODO use __contains__ here?
    def __iter__(self):
        return self.edges.iterkeys()

    def __str__(self):
        edges = ' '.join([e for e in self.edges])
        return 'Node(edges: {}, word: {})'.format(edges, self.word)

class Trie(object):

    def __init__(self, strings=[]):
        self.root = Node()
        for string in strings:
            self.insert(string)

    def insert(self, string):
        node = self.root
        for char in string:
            if char not in node:
                node[char] = Node()
            node = node[char]
        node.word = string

    def find(self, string):
        """Return node for last letter in string"""
        node = self.root
        for char in string:
            if char not in node:
                return False
            node = node[char]
        return node

    def has(self, string):
        """Return true if word is in trie"""
        node = self.find(string)
        if not node or not node.word:
            return False
        return True

    def remove(self, string):
        """Remove given string from trie"""
        raise NotImplementedError()

    def words_below(self, substring):
        """Find and return all words below substring"""
        start = self.find(substring)
        stack = [start] if start else []
        words = []
        while stack:
            node = stack.pop()
            if node.word:
                words.append(node.word)
            for edge in node:
                stack.append(node[edge])
        return words

    def __str__(self):
        pass

File: trie.js
/* Javascript translation of Python trie code
 * Isaac Stead, May 2013
 */

/* Trie node object */
function Node() {
    this.edges = {};
    this.word = null;
}

/* The trie data structure and methods */
function Trie() {
    this.root = new Node();
}

Trie.prototype.insert = function(string) {
    var node = this.root;
    for (var i = 0; i < string.length; i++) {
        var c = string[i];
        if (!node.edges[c]) {
            node.edges[c] = new Node();
        }
        node = node.edges[c];
    }
    node.word = string;
}

Trie.prototype.find = function(string) {
    var node = this.root;
    for (var i = 0; i < string.length; i++) {
        var c = string[i];
        if (!node.edges[c]) {
            return false;
        }
        node = node.edges[c];
    }
    return node;
}

Trie.prototype.has = function(string) {
    var node = this.find(string);
    if (!node || !node.word) {
        return false;
    } else {
        return true;
    }
}

Trie.prototype.words_below = function(substring) {
    var start = this.find(substring);
    var words = [];
    var stack = [];
    if (start) {
        stack = [start];
    }
    while (stack.length > 0) {
        var node = stack.pop();
        if (node.word) {
            words.push(node.word);
        }
        for (var edge in node.edges) {
            stack.push(node.edges[edge]);
        }
    }
    return words;
}

/* Testing functions */

function test() {
    var words = ['crips', 'creole', 'crew', 'crass', 'cray'];
    var trie = new Trie();
    // Test insertion
    for (var i = 0; i < words.length; i++) {
        trie.insert(words[i]);
    }
    console.log('insert => OK?');
    // Test has method
    if (trie.has('crips')) {
        console.log('has => OK');
    }
    // Test words_below method
    if (trie.words_below('cra') == ['cray', 'crass']) { // Order matters! do some comparison
        console.log('words_below => OK');
    }
    return trie;
}
